JOSS paper (#39)
PhilBoileau/cvCovEst: paper/paper.md

authors:
  - name: Philippe Boileau
    orcid: 0000-0002-4850-2507
    affiliation: "1, 2"
  - name: Brian Collica
    orcid: 0000-0003-1127-2557
    affiliation: 3
  - name: Nima S. Hejazi
    orcid: 0000-0002-7127-2789
    affiliation: "1, 2"
  - name: Mark J. van der Laan
    orcid: 0000-0003-1432-5511
    affiliation: "4, 3, 2"
  - name: Sandrine Dudoit
    orcid: 0000-0002-6069-8629
    affiliation: "3, 4, 2"
affiliations:
  - name: Graduate Group in Biostatistics, University of California, Berkeley
    index: 1
  - name: Center for Computational Biology, University of California, Berkeley
    index: 2
  - name: Department of Statistics, University of California, Berkeley
    index: 3
  - name: Division of Biostatistics, School of Public Health, University of California, Berkeley
    index: 4
date: 11 March 2021
bibliography: paper.bib
---

# Summary

Covariance matrices play fundamental roles in myriad statistical procedures.
When the observations in a dataset far outnumber the features, asymptotic theory
and empirical evidence have demonstrated the sample covariance matrix to be the
optimal estimator of this parameter. This assertion does not hold when the

> number of observations is commensurate or smaller than the number of features.

Reviewer: ```suggestion
number of observations is commensurate with or smaller than the number of features.
```

Consequently, statisticians have derived many novel covariance matrix estimators
for the high-dimensional regime, often relying on additional assumptions about
the parameter's structural characteristics (e.g., sparsity). While these
estimators have greatly improved the ability to estimate covariance matrices in
high-dimensional settings, objectively selecting the best estimator from among
the many possible candidates remains a largely unaddressed challenge. The
`cvCovEst` package addresses this methodological gap through its implementation
of a cross-validated framework for covariance matrix estimator selection. This

> data-adaptive procedure's selections are asymptotically optimal under minimal
> assumptions, meaning they are equivalent to the selections one would make if the
> true data-generating processes were known [@vdl2003unified].

Reviewer: ```suggestion
data-adaptive procedure's selections are asymptotically optimal under minimal
assumptions -- in fact, they are equivalent to the selections that would be made if given access to full knowledge of the true data-generating processes (i.e., an oracle selector) [@vdl2003unified].
```
The initial wording felt awkward since the second clause following "meaning" was not directly implied by the preceding clause.

performance is unsatisfactory: the sample covariance matrix is highly variable,
producing estimates with diverging condition numbers and over-dispersed
eigenvalues [@johnstone2001distribution]. These issues amplify estimation error
in resultant estimates, as well as in downstream analyses relying upon them.

As high-dimensional data have become widespread, researchers have derived many
novel covariance matrix estimators to remediate the sample covariance matrix's
deficiencies. These estimators come in many flavours, though most are
constructed by regularizing the sample covariance matrix. Comprehensive
reviews of these estimators are provided by @fan2016 and @pourahmadi2013.

This variety brings with it many challenges. Identifying an "optimal" estimator
from among a collection of candidates can prove a daunting task, one whose
objectivity is often compromised by the data analyst's decisions. Though
data-driven approaches for selecting an optimal estimator from among estimators
belonging to certain (limited) classes have been derived, the question of
selecting from a diverse collection of candidate procedures remains unaddressed.

# `cvCovEst` Framework

> The solution provided by `cvCovEst` a general, cross-validation-based,

Reviewer: ```suggestion
The solution provided by `cvCovEst` is a general, cross-validation-based,
```

estimator-agnostic framework for covariance matrix estimator selection. The
asymptotic optimality of selections are guaranteed under a few non-restrictive
assumptions by extending the seminal work of @laan_dudoit:2003, @dudoit2005, and
@vaart2006 on data-adaptive estimator selection to high-dimensional covariance
matrix estimation [@boileau2021]. Here, optimality is defined as choosing an
estimator with an equivalent risk difference to that which would have been

> selected were the underlying data-generating distribution _known_.

Reviewer: ```suggestion
selected were the underlying data-generating distribution _completely known_.
```

The `cvCovEst` software package implements this framework for the `R` language
and environment for statistical computing [@R]. Included is an accumulation of
covariance matrix estimators spanning the work of many researchers (Table 1).
They may be employed independently of the cross-validation procedure.
`cvCovEst` also provides a slew of plotting and summary functions. These
diagnostic tools allow users to gauge the algorithm's performance, diagnose
issues that might arise during estimation procedures, and build intuition about
the many estimators' behaviours. Additionally, users have options to increase
the cross-validation method's computational efficiency via parallel
computation. Parallelization relies on the suite of `future` packages
[@future] by way of the `origami` package [@origami].

> Table 1: Covariance matrix estimators implemented as of version 0.3.4.

Reviewer: ```suggestion
Table 1: Covariance matrix estimators implemented as of [version 0.3.4](https://github.com/PhilBoileau/cvCovEst).
```
We should eventually replace the link to the package with one specifying a git tag / GitHub release of v0.3.4

|Estimator | Implementation | Description |
|----------|----------|-------------|
| Sample covariance matrix | `sampleCovEst()` | The sample covariance matrix. |

> | Hard thresholding [@Bickel2008_thresh] | `thresholdingEst()` | Applies a hard thresholding operator to the entries of the sample covariance matrix |

Reviewer: ```suggestion
| Hard thresholding [@Bickel2008_thresh] | `thresholdingEst()` | Applies a hard thresholding operator to the entries of the sample covariance matrix. |
```
Insert "."

| SCAD thresholding [@rothman2009;@fan2001] | `scadEst()` | Applies the SCAD thresholding operator to the entries of the sample covariance matrix.|

> | Adaptive LASSO [@rothman2009] | `adaptiveLassoEst()` | Applies the adaptive LASSO thresholding operator to the entries of the sample covariance matrix |

Reviewer: ```suggestion
| Adaptive LASSO [@rothman2009] | `adaptiveLassoEst()` | Applies the adaptive LASSO thresholding operator to the entries of the sample covariance matrix. |
```
Insert "."

    )
    return(covmat * sign_mat)
}

# generate a 100 x 100 covariance matrix
sim_covmat <- toep_sim(p = 100, rho = 0.6, alpha = 0.3)

# sample 75 observations from multivariate normal mean = 0, var = sim_covmat
sim_dat <-  MASS::mvrnorm(n = 75, mu = rep(0, 100), Sigma = sim_covmat)

# run CV-selector
cv_cov_est_sim <- cvCovEst(
  dat = sim_dat,
  estimators = c(
    linearShrinkEst, thresholdingEst, bandingEst, taperingEst, sampleCovEst
  ),
  estimator_params = list(
    linearShrinkEst = list(alpha = seq(0.05, 0.25, 0.75)),
    thresholdingEst = list(gamma = seq(0.05, 0.25, 0.75)),
    bandingEst = list(k = seq(2L, 10L, 2L)),
    taperingEst = list(k = seq(2L, 10L, 2L))
  ),
  cv_scheme = "v_fold",
  v_folds = 5,
  center = TRUE
)

# plot a summary of the results
plot(cv_cov_est_sim, data_in = sim_dat)
```

![A summary of the `cvCovEst` procedure's results.](summary_plot.png){ width=80% }

# Availability

A stable release of the `cvCovEst` package is freely available via the
[Comprehensive `R` Archive Network](https://CRAN.R-project.org/package=cvCovEst).
Its development version can be found on
[GitHub](https://github.com/PhilBoileau/cvCovEst). Documentation and examples

> are contained in each version's manual pages and vignette.

Reviewer: ```suggestion
are contained in each version's manual pages, vignette, and `pkgdown` website (at https://philboileau.github.io/cvCovEst).
```

    index: 3

>   - name: Division of Epidemiology and Biostatistics, School of Public Health, University of California, Berkeley

Reviewer: ```suggestion
  - name: Division of Biostatistics, School of Public Health, University of California, Berkeley
```
I think the Biostatistics and Epidemiology Divisions have become independent after the merger a year or two ago.

    index: 4
date: 03 February 2021
bibliography: paper.bib
---

> 

Reviewer: High-level, the contents of the paper look good. Some general changes would be
* naming + citing exactly the covariance matrix estimators that are already implemented (stating the version number of the package for clarity)
* discussing more the visualization functionality, with a simple example. for example, we could grab the `README` example, mention the estimators that we select over, and then display one of a table and the nice plots that shows the performance within estimator families.
* in my most recent experience with submitting to JOSS, the editor asked that we re-organize the paper into short sections (as in https://github.com/nhejazi/txshift/blob/master/paper/paper.md). i think this is a good way to go, so that we have short paragraphs describing what `cvCovEst` does, what the problem is, and how we solve the problem (with technical details relegated to the methods paper)

Author: Done! See most recent comments.
